packages/context/test/unit/binding.unit.ts

```diff
@@ -173,6 +173,31 @@ describe('Binding', () => {
     });
   });
 
+  describe('apply(templateFn)', () => {
+    it('applies a template function', async () => {
+      binding.apply(b => {
+        b.inScope(BindingScope.SINGLETON).tag('myTag');
+      });
+      expect(binding.scope).to.eql(BindingScope.SINGLETON);
+      expect(binding.tagNames).to.eql(['myTag']);
+    });
+
+    it('sets up a placeholder value', async () => {
+      const toBeBound = (b: Binding<unknown>) => {
+        b.toDynamicValue(() => {
+          return Promise.reject(
```

Comment:
I feel that returning a rejected promise is suboptimal performance-wise. Have you considered throwing the error synchronously?

```ts
() => {
  throw new Error(...);
}
```